File: chd_auth.py
"""
长安大学统一身份认证平台 (IDS) 登录与会话管理
支持动态 Salt、AES-128-CBC 加密、验证码自适应检测
"""

import random
import base64
from urllib.parse import urljoin, quote
import requests
from bs4 import BeautifulSoup
from Crypto.Cipher import AES
from Crypto.Util.Padding import pad

AES_CHARS = "ABCDEFGHJKMNPQRSTWXYZabcdefhijkmnprstwxyz2345678"

def _random_string(length: int) -> str:
    """生成指定长度的随机字符串（与 CHD encrypt.js 字典一致）"""
    return ''.join(random.choice(AES_CHARS) for _ in range(length))

def encrypt_password(password: str, salt: str) -> str:
    """复现 CHD IDS 前端 encryptPassword 逻辑：
    AES-128-CBC, PKCS7 padding, 64位随机前缀，16位随机IV，Base64编码
    """
    salt = salt.strip()
    iv = _random_string(16)
    prefix = _random_string(64)
    data = (prefix + password).encode('utf-8')
    key = salt.encode('utf-8')
    iv_bytes = iv.encode('utf-8')
    
    cipher = AES.new(key, AES.MODE_CBC, iv_bytes)
    padded = pad(data, AES.block_size, style='pkcs7')
    encrypted = cipher.encrypt(padded)
    return base64.b64encode(encrypted).decode('utf-8')

class CHDAuthSession:
    def __init__(self):
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
                          '(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
            'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8'
        })
        self.service_url = 'http://bkjw.chd.edu.cn/eams/home.action'
        self.login_url = f'https://ids.chd.edu.cn/authserver/login?service={quote(self.service_url, safe="")}'

    def check_need_captcha(self, username: str) -> bool:
        """检测当前账号登录是否需要图形验证码"""
        try:
            url = f'https://ids.chd.edu.cn/authserver/checkNeedCaptcha.htl?username={username}'
            res = self.session.get(url, timeout=5)
            data = res.json()
            return bool(data.get('isNeed', False))
        except Exception:
            return False

    def is_session_alive(self) -> bool:
        """检查当前 session 是否仍然有效，避免频繁登录导致被校内风控锁定"""
        try:
            res = self.session.get(self.service_url, allow_redirects=False, timeout=5)
            if res.status_code == 200:
                return True
            loc = res.headers.get('Location', '')
            if res.status_code == 302 and 'authserver' not in loc:
                return True
            return False
        except Exception:
            return False

    def export_cookies(self) -> dict:
        """导出当前会话 Cookie 便于本地磁盘持久化"""
        return self.session.cookies.get_dict()

    def import_cookies(self, cookies: dict):
        """导入已有的 Cookie 恢复会话"""
        self.session.cookies.update(cookies)

    def login(self, username: str, password: str, captcha: str = "") -> dict:
        """执行全套统一身份认证与 EAMS 跳转登录
        返回: {"success": bool, "message": str, "cookies": dict}
        """
        try:
            # 1. 请求登录页面获取 Salt 与 Execution
            res = self.session.get(self.login_url, timeout=10)
            soup = BeautifulSoup(res.text, 'html.parser')

            pwd_form = soup.find('form', id='pwdFromId')
            if not pwd_form:
                # 可能已经处于登录状态
                if 'eams' in res.url:
                    return {"success": True, "message": "已在登录状态"}
                return {"success": False, "message": "无法定位登录表单，请稍后再试"}

            action = pwd_form.get('action') or '/authserver/login'
            if not action.startswith('http'):
                action = urljoin(res.url, action)
            if 'service=' not in action:
                action += f'?service={quote(self.service_url, safe="")}'

            salt_input = soup.find('input', id='pwdEncryptSalt')
            salt = salt_input.get('value') if salt_input else ''
            if not salt:
                return {"success": False, "message": "获取加密盐失败"}

            execution_input = pwd_form.find('input', attrs={'name': 'execution'})
            execution = execution_input.get('value') if execution_input else 'e1s1'

            # 2. 检查验证码要求
            need_captcha = self.check_need_captcha(username)
            if need_captcha and not captcha:
                return {"success": False, "need_captcha": True, "message": "需要输入验证码"}

            # 3. 加密密码
            enc_password = encrypt_password(password, salt)

            post_data = {
                'username': username,
                'password': enc_password,
                'captcha': captcha,
                '_eventId': 'submit',
                'cllt': 'userNameLogin',
                'dllt': 'generalLogin',
                'lt': '',
                'execution': execution
            }

            # 4. 提交登录
            login_res = self.session.post(action, data=post_data, allow_redirects=True, timeout=15)
            
            # 判断是否成功重定向至教务系统
            if 'bkjw.chd.edu.cn/eams' in login_res.url or 'home.action' in login_res.url:
                return {"success": True, "message": "登录成功"}
            
            # 解析失败提示信息
            err_soup = BeautifulSoup(login_res.text, 'html.parser')
            msg_el = err_soup.find(id='msg') or err_soup.find(class_='auth_error') or err_soup.find(class_='error-tips')
            err_msg = msg_el.get_text(strip=True) if msg_el else "学号或密码错误"
            return {"success": False, "message": err_msg}

        except Exception as e:
            return {"success": False, "message": f"登录请求异常: {str(e)}"}

File: chd_app/lib/services/chd_auth.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';
import 'package:dio/dio.dart';
import 'dart:io';
import 'package:cookie_jar/cookie_jar.dart';
import 'package:encrypt/encrypt.dart' as enc;
import 'package:html/parser.dart' as html_parser;

/// 安全 Cookie 管理拦截器，自动清洗和修复高校服务器非标准的 SameSite 属性，杜绝崩溃
class SafeCookieManager extends Interceptor {
  final CookieJar cookieJar;

  SafeCookieManager(this.cookieJar);

  static String sanitizeSetCookie(String raw) {
    return raw.replaceAllMapped(
      RegExp(r';\s*SameSite(=[^;]*)?', caseSensitive: false),
      (match) {
        final val = match.group(1)?.replaceFirst('=', '').trim().replaceAll('"', '') ?? '';
        final lower = val.toLowerCase();
        if (lower == 'lax') return '; SameSite=Lax';
        if (lower == 'strict') return '; SameSite=Strict';
        if (lower == 'none') return '; SameSite=None';
        return ''; // 移除其他非标准/空值的 SameSite
      },
    );
  }

  @override
  void onRequest(RequestOptions options, RequestInterceptorHandler handler) async {
    try {
      final cookies = await cookieJar.loadForRequest(options.uri);
      final cookieString = cookies.map((c) => '${c.name}=${c.value}').join('; ');
      if (cookieString.isNotEmpty) {
        options.headers[HttpHeaders.cookieHeader] = cookieString;
      }
    } catch (_) {}
    handler.next(options);
  }

  @override
  void onResponse(Response response, ResponseInterceptorHandler handler) async {
    await _saveCookies(response.realUri, response.headers);
    handler.next(response);
  }

  @override
  void onError(DioException err, ErrorInterceptorHandler handler) async {
    if (err.response != null) {
      await _saveCookies(err.response!.realUri, err.response!.headers);
    }
    handler.next(err);
  }

  Future<void> _saveCookies(Uri uri, Headers headers) async {
    final setCookies = headers[HttpHeaders.setCookieHeader];
    if (setCookies == null || setCookies.isEmpty) return;

    final parsedCookies = <Cookie>[];
    for (final raw in setCookies) {
      try {
        final sanitized = sanitizeSetCookie(raw);
        parsedCookies.add(Cookie.fromSetCookieValue(sanitized));
      } catch (_) {
        try {
          final parts = raw.split(';').first.split('=');
          if (parts.length >= 2) {
            parsedCookies.add(Cookie(parts[0].trim(), parts.sublist(1).join('=').trim()));
          }
        } catch (_) {}
      }
    }

    if (parsedCookies.isNotEmpty) {
      try {
        await cookieJar.saveFromResponse(uri, parsedCookies);
      } catch (_) {}
    }
  }
}

class AuthResult {
  final bool success;
  final String message;
  final bool needCaptcha;

  AuthResult({
    required this.success,
    this.message = '',
    this.needCaptcha = false,
  });
}

class CHDAuthService {
  late final Dio dio;
  late final CookieJar cookieJar;
  final String serviceUrl = 'http://bkjw.chd.edu.cn/eams/home.action';
  final String loginUrl =
      'https://ids.chd.edu.cn/authserver/login?service=http%3A%2F%2Fbkjw.chd.edu.cn%2Feams%2Fhome.action';

  CHDAuthService({CookieJar? customJar}) {
    cookieJar = customJar ?? CookieJar();
    dio = Dio(
      BaseOptions(
        connectTimeout: const Duration(seconds: 15),
        receiveTimeout: const Duration(seconds: 15),
        headers: {
          'User-Agent':
              'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
              '(KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36',
          'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8',
        },
        followRedirects: true,
        maxRedirects: 8,
        validateStatus: (status) => status != null && status < 400,
      ),
    );
    dio.interceptors.add(SafeCookieManager(cookieJar));
  }

  /// 检查当前 Cookie 会话是否依然有效，避免频繁登录导致账号被冻结 10 分钟
  Future<bool> isSessionAlive() async {
    try {
      final res = await dio.get(
        serviceUrl,
        options: Options(
          followRedirects: false,
          validateStatus: (status) => status != null && status < 500,
        ),
      );
      if (res.statusCode == 200) {
        return true;
      }
      final location = res.headers.value('location') ?? '';
      if (res.statusCode == 302 && !location.contains('authserver')) {
        return true;
      }
      return false;
    } catch (_) {
      return false;
    }
  }

  /// 检查是否需要验证码
  Future<bool> checkNeedCaptcha(String username) async {
    try {
      final res = await dio.get(
        'https://ids.chd.edu.cn/authserver/checkNeedCaptcha.htl?username=$username',
      );
      if (res.data is Map) {
        return res.data['isNeed'] == true;
      } else if (res.data is String) {
        final map = jsonDecode(res.data as String);
        return map['isNeed'] == true;
      }
      return false;
    } catch (_) {
      return false;
    }
  }

  /// 获取验证码二进制流
  Future<Uint8List?> getCaptchaBytes() async {
    try {
      final rnd = Random().nextInt(1000000);
      final res = await dio.get<List<int>>(
        'https://ids.chd.edu.cn/authserver/getCaptcha.htl?$rnd',
        options: Options(responseType: ResponseType.bytes),
      );
      if (res.data != null) {
        return Uint8List.fromList(res.data!);
      }
    } catch (_) {}
    return null;
  }

  /// 纯 Dart 实现 AES-128-CBC 密码加密
  String _encryptPassword(String password, String salt) {
    const chars = 'ABCDEFGHJKMNPQRSTWXYZabcdefhijkmnprstwxyz2345678';
    final rnd = Random();
    final random64 =
        List.generate(64, (_) => chars[rnd.nextInt(chars.length)]).join();
    final iv16 =
        List.generate(16, (_) => chars[rnd.nextInt(chars.length)]).join();

    final key = enc.Key.fromUtf8(salt);
    final iv = enc.IV.fromUtf8(iv16);
    final encrypter = enc.Encrypter(
      enc.AES(key, mode: enc.AESMode.cbc, padding: 'PKCS7'),
    );

    final encrypted = encrypter.encrypt(random64 + password, iv: iv);
    return encrypted.base64;
  }

  /// 执行登录统一认证
  Future<AuthResult> login(
    String username,
    String password, {
    String captcha = '',
  }) async {
    try {
      // 1. 请求登录页面，获取 salt 和 execution
      final getRes = await dio.get(loginUrl);
      final doc = html_parser.parse(getRes.data);

      final pwdForm = doc.querySelector('form#pwdFromId');
      if (pwdForm == null) {
        if (getRes.realUri.toString().contains('eams')) {
          return AuthResult(success: true, message: '已经在登录状态');
        }
        return AuthResult(success: false, message: '未找到登录表单，请检查网络');
      }

      var action = pwdForm.attributes['action'] ?? '/authserver/login';
      if (!action.startsWith('http')) {
        action = Uri.parse(loginUrl).resolve(action).toString();
      }
      if (!action.contains('service=')) {
        action +=
            '?service=${Uri.encodeQueryComponent(serviceUrl)}';
      }

      final saltInput = doc.querySelector('input#pwdEncryptSalt');
      final salt = saltInput?.attributes['value'] ?? '';
      if (salt.isEmpty) {
        return AuthResult(success: false, message: '获取动态密码盐失败');
      }

      final executionInput = doc.querySelector('input[name=execution]');
      final execution = executionInput?.attributes['value'] ?? 'e1s1';

      // 2. 检查验证码
      final needCaptcha = await checkNeedCaptcha(username);
      if (needCaptcha && captcha.isEmpty) {
        return AuthResult(
          success: false,
          needCaptcha: true,
          message: '需要输入验证码',
        );
      }

      // 3. 加密密码
      final encPwd = _encryptPassword(password, salt);

      // 4. 发送登录请求
      final formData = {
        'username': username,
        'password': encPwd,
        'captcha': captcha,
        '_eventId': 'submit',
        'cllt': 'userNameLogin',
        'dllt': 'generalLogin',
        'lt': '',
        'execution': execution,
      };

      final postRes = await dio.post(
        action,
        data: formData,
        options: Options(
          contentType: Headers.formUrlEncodedContentType,
          followRedirects: true,
          validateStatus: (status) => status != null && status < 500,
        ),
      );

      // 手动逐步跟随跨协议重定向，确保每一步 302 返回的 Set-Cookie 均被保存
      var currentUrl = postRes.headers.value('location') ?? '';
      var redirectSteps = 0;
      while (currentUrl.isNotEmpty && redirectSteps < 10) {
        redirectSteps++;
        if (!currentUrl.startsWith('http')) {
          currentUrl = Uri.parse(action).resolve(currentUrl).toString();
        }
        final stepRes = await dio.get(
          currentUrl,
          options: Options(
            followRedirects: false,
            validateStatus: (status) => status != null && status < 500,
          ),
        );
        final nextLoc = stepRes.headers.value('location') ?? '';
        if (nextLoc.isNotEmpty) {
          if (nextLoc.startsWith('http')) {
            currentUrl = nextLoc;
          } else {
            currentUrl = Uri.parse(currentUrl).resolve(nextLoc).toString();
          }
        } else {
          break;
        }
      }

      // 验证教务会话是否已激活
      if (await isSessionAlive()) {
        return AuthResult(success: true, message: '登录成功');
      }

      final finalUrl = postRes.realUri.toString();
      if (finalUrl.contains('eams') || finalUrl.contains('home.action')) {
        return AuthResult(success: true, message: '登录成功');
      }

      // 登录失败，从页面提取错误提示
      final errDoc = html_parser.parse(postRes.data);
      final errEl = errDoc.querySelector('#showErrorTip, #msg, .auth_error');
      final errMsg = errEl?.text.trim() ?? '';
      if (errMsg.isNotEmpty) {
        return AuthResult(
          success: false,
          message: errMsg,
          needCaptcha: needCaptcha,
        );
      }

      if (postRes.data.toString().contains('密码有误') ||
          postRes.data.toString().contains('用户名或密码')) {
        return AuthResult(success: false, message: '用户名或密码错误');
      }

      return AuthResult(success: true, message: '登录完成');
    } catch (e) {
      return AuthResult(success: false, message: '连接错误: $e');
    }
  }
}
